heudiconv/bids.py

```diff
@@ -487,3 +510,505 @@ def convert_sid_bids(subject_id):
+        # We want to force the matching, so just return some string
+        # regardless of the image
+        key_info = [KeyInfoForForce]
+
+    return key_info
+
+
+def find_compatible_fmaps_for_run(json_file, fmap_groups, matching_parameters):
+    """
+    Finds compatible fmaps for a given run, for populate_intended_for.
+
+    Parameters:
+    ----------
+    json_file : str or os.path
+        path to the json file
+    fmap_groups : dict
+        key: prefix common to the group
+        value: list of all fmap paths in the group
+    matching_parameters : list of str from AllowedFmapParameterMatching
+        matching_parameters that will be used to match runs
+
+    Returns:
+    -------
+    compatible_fmap_groups : dict
+        Subset of the fmap_groups which match json_file, according
+        to the matching_parameters.
+        key: prefix common to the group
+        value: list of all fmap paths in the group
+    """
+    if type(matching_parameters) is not list:
+        matching_parameters = [matching_parameters]
+
+    lgr.debug('Looking for fmaps for %s', json_file)
+    json_info = {}
+    for param in matching_parameters:
+        if param not in AllowedFmapParameterMatching:
+            raise ValueError(
+                "Fmap matching_parameter %s not allowed." % param
+            )
+        json_info[param] = get_key_info_for_fmap_assignment(json_file, param)
+
+    compatible_fmap_groups = {}
+    for fm_key, fm_group in fmap_groups.items():
+        # check the key_info (for all parameters) for one (the first) of
+        # the fmaps in the group:
+        compatible = False
+        for param in matching_parameters:
+            fm_info = get_key_info_for_fmap_assignment(fm_group[0], param)
+            # for the case in which key_info is a list of strings:
+            if type(json_info[param][0]):
+                compatible = json_info[param] == fm_info
+            else:
+                # allow for tiny differences between the affines etc
+                compatible = all(np.allclose(x, y) for x, y in zip(json_info[param], fm_info))
+            if not compatible:
+                continue     # don't bother checking more params
+        if compatible:
+            compatible_fmap_groups[fm_key] = fm_group
+
+    return compatible_fmap_groups
+
+
+def find_compatible_fmaps_for_session(path_to_bids_session, matching_parameters):
+    """
+    Finds compatible fmaps for all non-fmap runs in a session.
+
+    Parameters:
+    ----------
+    path_to_bids_session : str or os.path
+        path to the session folder (or to the subject folder, if there are no
+        sessions).
+    matching_parameters : list of str from AllowedFmapParameterMatching
+        matching_parameters that will be used to match runs
+
+    Returns:
+    -------
+    compatible_fmap : dict
+        Dict of compatible_fmaps_groups (values) for each non-fmap run (keys)
+    """
+    if type(matching_parameters) is not list:
```

Comment:
```suggestion
    if isinstance(matching_parameters, list):
        assert isinstance(matching_parameters, str), "matching_parameters must be a str or a list, got %s" % matching_parameters
```
note: didn't check if would work as expected

Reply:
Shouldn't it be `if not isinstance(matching_parameters, list):`? (Note the `not`)

Reply:
right, sorry for a buggy suggestion -- fixed it now!